src/transforms/projectionpursuit.jl

```diff
@@ -0,0 +1,203 @@
+randomly generated samples from the standard multivariate Gaussian distribution, or when the 
+number of iterations reaches a maximum `maxiter`.
+
+# Examples
+
+```julia
+ProjectionPursuit()
+ProjectionPursuit(deg=10)
+ProjectionPursuit(perc=.85, n=50)
+ProjectionPursuit(tol=1e-4, maxiter=250, deg=5, perc=.95, n=100)
+```
+
+"""
+
+struct ProjectionPursuit{T} <: Stateless
+  tol::T
+  maxiter::Int
+  deg::Int
+  perc::T
+  n::Int
+end
+
+ProjectionPursuit(;tol=1e-6, maxiter=100, deg=5, perc=.95, n=100) =
+  ProjectionPursuit{typeof(tol)}(tol, maxiter, deg, perc, n)
+
+isrevertible(::Type{<:ProjectionPursuit}) = true
+
+# transforms a row of random variables into a convex combination 
+# of random variables with values in [-1,1] and standard normal distribution
+function rscore(Z, α)
+  ᾱ = (1/norm(α)) .* α
+  X = Z * ᾱ
+  2 .* cdf.(Normal(), X) .- 1
+end
+
+# projection index of sample along a given direction
+function pindex(transform, Z, α)
+  d = transform.deg
+  X = rscore(Z, α)
+  I = (3/2) * mean(X)^2
+  if d > 1
+    Pⱼ₋₂, Pⱼ₋₁ = ones(size(X)), X
+    for j = 2:d
+      Pⱼ₋₂, Pⱼ₋₁ = 
+        Pⱼ₋₁, (1/j) * ((2j-1) * X .* Pⱼ₋₁ - (j-1) * Pⱼ₋₂)
+      I += ((2j+1)/2) * (mean(Pⱼ₋₁))^2
+    end
+  end
+  I
+end
+
+# projection index of the standard multivariate Gaussian
+function gindex(transform, N, q)
+  n  = transform.n
+  E  = 1.0 * I(q)
+  Iₙ = Matrix{Float64}(undef, n, q)
+  for i = 1:n
+    S = randn(N, q)
+    for j = 1:q
+      Iₙ[i,j] = pindex(transform, S, E[:,j])
+    end
+  end
+  Iₙ
+end
+
+function optimaxis(transform, Z)
+  _, q = size(Z)
+  E = 1.0 * I(q)
+  f(i) = pindex(transform, Z, E[:,i])
+  iₘₐₓ = argmax(f, 1:q)
+  E[:,iₘₐₓ]
+end
+
+function optimdiag(transform, Z, α₀)
+  q  = length(α₀)
+  E  = 1.0 * I(q)
+  
+  # initialize score and direction
+  Iₐ = pindex(transform, Z, α₀)
+  α  = α₀
+  
+  # evaluate score along diagonals
+  diag(α, s, e) = (1/√(2+2s*α[i])) * (α + s*e)
+  func(d) = pindex(transform, Z, d)
+  for eᵢ in eachcol(E)
+    d₊ = diag(α, +1, eᵢ)
+    d₋ = diag(α, -1, eᵢ)
+    f₊ = func(d₊)
+    f₋ = α[i] != 1 ? func(d₋) : 0
+    f, d = f₊ > f₋ ? (f₊, d₊) : (f₋, d₋)
+    if f > Iₐ
+      Iₐ = f
+      α  = d
+    end
+  end
+  
+  α
+end
+
+function optimall(transform, Z, α₀)
+  f(α) = -pindex(transform, Z, α)
+  op = optimize(f, α₀)
+  minimizer(op)
+end
+
+function alphamax(transform, Z)
+  α₁ = optimaxis(transform, Z)
+  α₂ = optimdiag(transform, Z, α₁)
+  optimall(transform, Z, α₂)  
+end
+
+function rmstructure(transform, Z, α)
+  q = length(α)
+  Q, R = qr(hcat(α, rand(q, q-1)))
+  while norm(diag(R)) < transform.tol
+    Q, R = qr(hcat(α, rand(q, q-1)))
+  end
+  T = Z * Q
+  table = Tables.table(T)
+  newtable, colcache = apply(Quantile(), table)
+  T[:,1] = Tables.getcolumn(newtable, 1)
+  T * Q', Q, colcache
+end
+
+function stopcriteria(transform, Z, qgauss)
+  _, q = size(Z)
+  E = 1.0 * I(q)
+  Iₐ = [pindex(transform, Z, col) for col in eachcol(E)]
+  !(qgauss > Iₐ)
```

Comment:
Notice that `qgauss` and `Iₐ` are both vectors. The syntax `(qgauss > Iₐ)` doesn't mean what you think it means:

```julia
julia> a = [1,2,3]
3-element Vector{Int64}:
 1
 2
 3

julia> b = [4,1,2]
3-element Vector{Int64}:
 4
 1
 2

julia> a < b
true
```

I think we want `any(Iₐ .> qgauss)` right?

Reply:
@DianaPat this issue is still valid, correct?